1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,279 @@
+    }
+    process {
+        $Ping.Send($Target, $Timeout, $Buffer, $PingOptions)
+    }
+    cleanup {
+        if ($Ping) {
+            $Ping.Dispose()
+        }
+    }
+}
+
+'1.2.3.4', 'www.google.com', '8.8.8.8' | Get-PingReply
+```
+
+#### Write to a File Using .NET Streams
+
+```powershell
+using namespace System.IO
+
+function Write-File {
+    [CmdletBinding(PositionalBinding = $false)]
+    param(
+        [Parameter(ValueFromPipeline)]
+        [string] $InputObject,
+
+        [Parameter(Position = 0, Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $Path
+    )
+    begin {
+        $fileStream = [FileStream]::new(
+            <# path:   #> $Path,
+            <# mode:   #> [FileMode]::Open,
+            <# access: #> [FileAccess]::Write,
+            <# share:  #> [FileShare]::Read)
+
+        $streamWriter = [StreamWriter]::new($fileStream)
+    }
+    process {
+        if ([string]::IsNullOrWhiteSpace($InputObject)) {
+            return
+        }
+
+        $streamWriter.WriteLine($InputObject)
+    }
+    end {
+        $StreamWriter.Flush()
+    }
+    cleanup {
+        if ($streamWriter) {
+            $streamWriter.Dispose()
+        }
+
+        if ($fileStream) {
+            $fileStream.Dispose()
+        }
+    }
+}
+
+"Text" | Write-File -Path "C:\tmp.txt"
+```
+
+### Considerations
+
+#### Breaking Change
+
+This proposal is a breaking change, as it prevents users from directly calling a function, alias, or command named `cleanup` (since this will now parse as a keyword and not a command name).
+However, such commands may still be invoked with the call operator (`& cleanup`).
+
+It also prevents users from defining their own `cleanup` dynamic keyword, though use of dynamic keywords is uncommon, so this is unlikely to become a concern.
+
+#### Behaviour with Ctrl+C During `Cleanup {}`
+
+As PowerShell is primarily an administrative shell, it _typically_ (though not always) respects `Ctrl+C` and halts processing.
+Due to the nature and purpose of the `cleanup {}` block, it's not suitable to permit `cleanup {}` blocks to be cancelled during their execution.
+Allowing `cleanup {}` blocks to be cancelled may lead to memory leaks and other issues that can potentially arise from preventing proper disposal of `IDisposable` resources.
+
+However, this means that actions taken within `cleanup {}` can potentially be disruptive, and it is the author's responsibility to ensure their `cleanup {}` actions do not indefinitely hang.
+Compiled cmdlets may already implement `IDisposable` without restriction to similar effect, it is unlikely we have a pressing need to deviate from this behaviour.
+Customer feedback may be necessary to determine if taking action towards this course is necessary.
+
+To give a user appropriate feedback, we could emit a warning message that is displayed immediately when a pipeline stop is requested, before any command starts its dispose routines:
```

Comment:
could you suggest what you are planning here rather than leaving ambiguous? Are you planning to provide a warning?